### PR TITLE
feat: dynamic api reference

### PR DESCRIPTION
**Pull request by manika-signoz**
## Summary

Replaces the checked-in OpenAPI bundle with a **version-aware API reference** that loads `docs/api/openapi.yml` from the **SigNoz/signoz** repo per release tag, discovers versions from the **GitHub Releases** API, and adds routing, caching, bot/LLM-friendly YAML rewrites, and UI for version selection.

Closes https://github.com/SigNoz/platform-pod/issues/2119

## Motivation / Problem

The site previously shipped a single static `public/openAPISpec/api.yaml`, which drifted from product releases and did not let readers pick a SigNoz version. Serving specs from the same tags as releases keeps the reference accurate and discoverable without manual asset updates.

## Changes

- Remove bundled `public/openAPISpec/api.yaml` and fetch specs from `https://raw.githubusercontent.com/SigNoz/signoz/{version}/docs/api/openapi.yml` with 24h revalidation.
- Add `utils/apiReference.ts`: list releases via GitHub API (paginated), filter to semver ≥ `0.105.0`, exclude drafts/prereleases, keep highest patch per minor line, sort newest first; optional `GITHUB_TOKEN` / `GITHUB_TOKEN` usage for rate limits; `unstable_cache` for the version list.
- Add `app/api-reference/page.tsx`: index page listing available versions with “Latest” on the first entry.
- Add `app/api-reference/[version]/page.tsx`: `latest` redirects to resolved latest tag; validates semver; loads spec + versions; `force-dynamic` for that segment; per-version metadata.
- Add `components/APIVersionSwitcher` for switching between loaded versions.
- Add `app/api/api-reference-openapi/[version]/route.ts`: serves raw YAML with CDN-friendly `Cache-Control` for consumers and middleware rewrites.
- Add `utils/apiReferenceMarkdownRouting.ts` + **middleware**: bots requesting `Accept: text/markdown` get rewritten to the OpenAPI YAML route (aligned with docs markdown rewrite behavior).
- Update `components/OpenAPISpec`: SSR placeholder uses static stacked layout and `hideTryIt`; client uses hash router and responsive layout after mount (avoids dynamic server usage / hydration issues).
- Add `constants/apiReference.ts` for URLs, minimum version, revalidate interval, and cache keys/tags.
- `utils/getPageType.ts`: classify `/api-reference` paths for analytics.

## Description

Visitors open `/api-reference`, pick a version (or use `/api-reference/latest/`, which redirects to the newest eligible release tag), and see Stoplight Elements backed by the YAML for that tag. Specs and the release list are cached and revalidated on a 24-hour cadence to balance freshness and GitHub load.

## Type of Change

- [ ] **Docs** – Changes to `data/docs/**`
- [ ] **Blog** – Changes to `data/blog/**`
- [x] **Site Code** – Changes to `app/**`, `components/**`, `hooks/**`, `utils/**`, config, etc.
- [ ] **Redirects** – Renamed/moved docs or updated `next.config.js` redirects
- [ ] **Dependencies / CI / Scripts**
- [ ] **Other** – Explain below

## Impact

- [ ] Documentation only
- [x] UI changes
- [ ] Developer workflow
- [ ] Performance impact
- [ ] Breaking change

If breaking change, describe migration steps:

## Context & Screenshots
TBA

## Testing
- [ ] Tested locally
- [ ] Edge cases considered
- [ ] Existing functionality verified
- [ ] New tests added (if applicable)

Steps to test:
1. Open `/api-reference` and confirm the version list matches expectations (ordering, “Latest” badge on first row).
2. Open `/api-reference/latest/` and confirm redirect to a concrete semver tag.
3. Open `/api-reference/<valid-tag>/` and confirm Stoplight renders; try the version switcher.
4. Request `/api-reference/<version>/` with a bot-like UA and `Accept: text/markdown` (non-production: check `x-markdown-rewrite` if enabled) and confirm YAML is served.
5. Run `yarn build` and `yarn lint` on the branch before marking ready.

## Rollout / Deployment Notes

- **Optional:** Configure `GITHUB_TOKEN` in the deployment environment to reduce GitHub API rate-limit risk when resolving the version list (raw spec fetch can also use token where implemented).
- Ensure outbound access to `api.github.com` and `raw.githubusercontent.com` is allowed from the deployment runtime.

## Additional Context

- Vercel preview: see the latest deployment comment on the PR (e.g. [PR #3181](https://github.com/SigNoz/signoz.io/pull/3181)).
- “Try it” is hidden in Stoplight per current product choice (`hideTryIt`).

---

**Note:** Submit as Draft by default. Mark "Ready for review" when checks pass and content is ready.

### CODE REVIEW

File: utils/apiReference.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  if (!parts) return null
+
+  const url = GITHUB_SPEC_RAW_URL_TEMPLATE.replace('{version}', version)
+  try {
+    const res = await fetch(url, {
+      headers: {
+        'User-Agent': 'signoz-api-reference',
+        ...(process.env.GITHUB_TOKEN ? { Authorization: `token ${process.env.GITHUB_TOKEN}` } : {}),
+      },
+      next: { revalidate: API_SPEC_REVALIDATE_SECONDS },
```
[the remark this answers is not in the document]
Comment (manika-signoz) [reply utils/apiReference.ts:62]: done

File: app/api-reference/[version]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import {
+  fetchOpenAPISpec,
+  fetchAvailableAPIVersions,
+  resolveLatestVersion,
+} from '@/utils/apiReference'
+import siteMetadata from '@/data/siteMetadata'
+import OpenAPISpec from '@/components/OpenAPISpec'
+import APIVersionSwitcher from '@/components/APIVersionSwitcher'
+
+export const dynamic = 'force-dynamic'
```
[the remark this answers is not in the document]
Comment (manika-signoz) [reply app/api-reference/[version]/page.tsx]: done

File: components/OpenAPISpec/index.tsx [hunk 3]
```diff
@@ -3,8 +3,25 @@
...
+  useEffect(() => {
+    setMounted(true)
+  }, [])
+
+  if (!mounted)
+    return (
+      <div className="hidden h-screen w-screen">
+        <API apiDescriptionDocument={specContent} router="static" layout="stacked" hideTryIt />
+      </div>
+    )
```
[the remark this answers is not in the document]
Comment (manika-signoz) [reply components/OpenAPISpec/index.tsx:27]: needed for SEO indexing of individual endpoints

File: components/APIVersionSwitcher/index.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+  useEffect(() => {
+    setPortalContainer(null)
+    document.querySelectorAll('.api-version-portal').forEach((el) => el.remove())
+
+    if (findAndAttach()) return undefined
+
+    const interval = setInterval(() => {
+      if (findAndAttach()) clearInterval(interval)
+    }, 200)
```
[the remark this answers is not in the document]
Comment (manika-signoz) [reply components/APIVersionSwitcher/index.tsx:120]: done

File: app/api/api-reference-openapi/[version]/route.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    status: 404,
+    headers: {
+      'Content-Type': 'text/plain; charset=utf-8',
+      'Cache-Control': 'no-store',
+    },
+  })
+
+export async function GET(_: Request, { params }: { params: { version: string } }) {
+  const raw = params.version
+  const version = raw === 'latest' ? await resolveLatestVersion() : decodeURIComponent(raw)
```
[the remark this answers is not in the document]
Comment (manika-signoz) [reply app/api/api-reference-openapi/[version]/route.ts]: used `const version = raw === 'latest' ? await resolveLatestVersion() : raw`

File: utils/apiReference.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  prerelease: boolean
+}
+
+function githubFetchHeaders(): Record<string, string> {
+  const headers: Record<string, string> = {
+    Accept: 'application/vnd.github+json',
+    'User-Agent': 'signoz-api-reference',
+  }
+  if (process.env.GITHUB_TOKEN) {
+    headers.Authorization = `Bearer ${process.env.GITHUB_TOKEN}`
```
Comment (H4ad) [inline utils/apiReference.ts:33]: When creating this token, ensure we use less permissions as possible.

File: utils/apiReference.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    return null
+  }
+}
+
+export interface APIVersionInfo {
+  version: string
+  publishedAt: string
+}
+
+async function fetchAllAPIVersions(): Promise<APIVersionInfo[]> {
```
Comment (H4ad) [inline utils/apiReference.ts:80]: In future, we can create an automation that fetches this and commit to the repo, like a snapshot, so in the best case we could just fetch one page.

Not sure if this cache is invalidated after release, so that's why I suggested this.

**Review by H4ad — APPROVED** [review]

File: utils/apiReference.ts (hunk 6 above)
Comment (manika-signoz) [reply utils/apiReference.ts:33]: yes, currently we also store this call in cache but this is only added for future, in case we run into github rate limits

File: utils/apiReference.ts (hunk 7 above)
Comment (manika-signoz) [reply utils/apiReference.ts:80]: the api has a 24h ttl, so it would be revalidated automatically and we can also do an on demand revalidation if needed
